src/main/scala/chisel3/simulator/ChiselSimAPI.scala

```diff
@@ -0,0 +1,458 @@
+          message match {
+            case Some(message) =>
+              expectLeaf(s, d, message)(sourceInfo)
+            case _ =>
+              expectLeaf(s, d)(sourceInfo)
+          }
+        case _ =>
+          throw new Exception("type not supported!")
+      }
+    }
+  }
+
+  implicit final class testableValidIO[T <: Data](sig: ValidIO[T])(implicit sourceInfo: SourceInfo)
+      extends clockedInterface {
+
+    def enqueue(data: T) = {
+      require(data.isLit, "enqueued data must be literal!")
+      sig.valid.poke(true)
+      pokeRec(sig.bits, data)
+      clock.step()
+      sig.valid.poke(false)
+    }
+
+    def enqueueSeq(dataSeq: Seq[T]) = {
+      for (data <- dataSeq) {
+        enqueue(data)
+      }
+    }
+
+    def waitForValid() = {
+      // println("wait for valid")
+      // clock.stepUntil(sig.valid, 1, maxWaitCycles)
+      // val timeout = sig.valid.peekValue().asBigInt == 0
+      // chisel3.assert(!timeout, s"Timeout after $maxWaitCycles cycles waiting for valid")
+      var cycles = 0
+      while (!sig.valid.peek().litToBoolean) {
+        clock.step()
+        Predef.assert(cycles < maxWaitCycles, "Timeout after ${maxWaitCycles} cycles waiting for valid")
+        cycles += 1
+      }
+    }
+
+    def dequeue(): HierarchicalValue = {
+      waitForValid()
+      val value = peekHierValueRec(sig)
+      clock.step()
+      value
+    }
+
+    def expectDequeue(expected: T, message: String): Unit =
+      expectDequeue(expected, Some(message))
+
+    def expectDequeue(expected: T, message: Option[String] = None): Unit = {
+      require(expected.isLit, "expected value must be a literal!")
+      waitForValid()
+      expectRec(sig.bits, expected, message)
+      clock.step()
+    }
+
+    def expectDequeueSeq(dataSeq: Seq[T], message: String): Unit = {
+      for ((data, i) <- dataSeq.zipWithIndex) {
+        expectDequeue(data, message)
+      }
+    }
+
+    def expectDequeueSeq(dataSeq: Seq[T]): Unit = {
+      for ((exp, i) <- dataSeq.zipWithIndex) {
+        expectDequeue(exp, s"Element $i was different from expected value: $exp!")
+      }
+    }
+
+  }
+
+  implicit final class testableDecoupledIO[T <: Data](sig: DecoupledIO[T])(implicit sourceInfo: SourceInfo)
+      extends clockedInterface {
+
+    def enqueue(data: T) = {
+      require(data.isLit, "enqueued data must be literal!")
+      sig.valid.poke(true)
+      pokeRec(sig.bits, data)
+      waitForReady()
+      clock.step()
+      sig.valid.poke(false)
+    }
+
+    def enqueueSeq(dataSeq: Seq[T]) = {
+      for (data <- dataSeq) {
+        enqueue(data)
+      }
+    }
+
+    def waitForValid() = {
+      // println("wait for valid")
+      // clock.stepUntil(sig.valid, 1, maxWaitCycles)
+      // val timeout = sig.valid.peekValue().asBigInt == 0
+      // chisel3.assert(!timeout, s"Timeout after $maxWaitCycles cycles waiting for valid")
+      var cycles = 0
+      while (!sig.valid.peek().litToBoolean) {
+        clock.step()
```

Comment:
This would be a great place to take advantage of SVSims clock ticking with sentinel values: https://github.com/chipsalliance/chisel/blob/2fdaead511d37e1b1c0badbb2a4077506b538bb5/svsim/src/test/scala/BackendSpec.scala#L151 this allows for decoupling to offset the IPC overhead.